sensor: eeprom: Add TMP116 EEPROM access

kurddt opened the pull request:
Read and write accesses are supported by using
either custom or eeprom API

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

Commit:
sensor: eeprom: Add TMP116 EEPROM access

eeprom access is supported by using
either custom or eeprom API

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

henrikbrixandersen commented:
> Only read-only access is supported by using either custom or eeprom API

Thank you. How come only read-only access is implemented?

kurddt commented:
Mainly because it's the only thing I used and that I tested. It allows to read the IDs preprogrammed in the sensor eeprom.

If I'm able clear some time I will add the write part, but I hope this will not block the PR

kurddt commented on drivers/sensor/tmp116/tmp116.c:
```diff
@@ -46,6 +51,95 @@ static int tmp116_reg_write(const struct device *dev, uint8_t reg,
+	}
+
+	/* TMP117 uses EEPROM[2] as temperature offset register */
+	if (drv_data->id == TMP117_DEVICE_ID &&
```
I'm not sure about this part. Is it better to do as if the TMP117 have a contiguous EEPROM of 3*16bits instead of 4*16bits with reserved 16bits

kurddt commented on drivers/sensor/tmp116/Kconfig:
```diff
@@ -8,3 +8,12 @@ config TMP116
+		default SENSOR_INIT_PRIORITY
+		help
+		  Device driver initialization priority.
+endif
```
For my own understanding what's the difference between the two construction and when it is advised to use one or the other ?

henrikbrixandersen replied:
The recommendation is to use `depends on` if it's just a single option, but an `if` block if multiple options depend on the same option.

kurddt commented on include/drivers/sensor/tmp116.h:
```diff
@@ -0,0 +1,21 @@
+#define ZEPHYR_INCLUDE_DRIVERS_SENSOR_TMP116_H_
+
+#include <device.h>
+#include <drivers/eeprom.h>
```
Only for `off_t` I can change it to `sys/types.h` if you prefer

henrikbrixandersen replied:
Please do.

Commit:
samples: tmp116: Add access to eeprom

Add read access to internal tmp116 eeprom which
contains the device unique id after manufacturing

Signed-off-by: Guillaume Lager <g.lager@innoseis.com>

henrikbrixandersen reviewed (approved):
Looks good, thank you. One minor nit, not a blocker.

henrikbrixandersen commented:
@MaureenHelm Are you ok with the changes to the sensor driver?

MaureenHelm commented on drivers/sensor/tmp116/Kconfig:
```diff
@@ -8,3 +8,10 @@ config TMP116
 	help
 	  Enable driver for TMP116 and TMP117 temperature sensors.
+
+	config SENSOR_TMP116_INIT_PRIORITY
```
Is this really needed? I've been working on reducing the number of driver-specific init priority configs and instead using common configs for each driver class (e.g., SENSOR_INIT_PRIORITY, EEPROM_INIT_PRIORITY, etc.)

kurddt replied:
out of curiosity what is the reason for limiting those driver-specific priorities ? 
Usually for my applications I end up with a lot of driver-specific init priority defined otherwise because of devices dependency some initialization fails. That said it's probably less true for "leaf" devices like sensors or eeprom. I will review the usage.

In any case I will change the default value of `EEPROM_TMP116_INIT_PRIORITY` to `EEPROM_INIT_PRIORITY`

kurddt replied:
I think this config is not really required in my application so I will remove it as per your suggestion. `EEPROM_INIT_PRIORITY` doesn't seem to exist so it's propably why I set `EEPROM_TMP116_INIT_PRIORITY` default to an arbitrary value

MaureenHelm replied:
Please see #40044 which adds `EEPROM_INIT_PRIO`.

This cleanup of initialization priorities was triggered by #38631. Many drivers use `KERNEL_INIT_PRIORITY_DEVICE` directly, which can cause problems if you need to initialize two drivers in a particular order, in that case it was a clock driver and a UART driver. We needed a little more granularity to specify initialization priorities, but didn't want to introduce the Kconfig symbol sprawl that would come with specifying priorities for every single driver. It will also make applications more portable across multiple boards.

kurddt replied:
@MaureenHelm I rebased on main after  #40044 was merged and removed the driver specific init priority in favor of `EEPROM_INIT_PRIORITY`

kurddt commented:
@MaureenHelm Could you review again ?
